python: hint dependencies

tdsmith opened the pull request:
Add include paths to CPPFLAGS and library paths to LDFLAGS so extension modules find the Homebrew libraries on which they depend.

Fixes #420.

- [x] Test on Trusty
- [x] Test on CentOS 7
- [x] Apply symmetric change to python3.rb
- [x] Check whether `-I#{HOMEBREW_PREFIX}/include -L#{HOMEBREW_PREFIX}/lib` is sufficient

tdsmith commented:
- [x] Test on Trusty
- [ ] Test on CentOS 7
- [ ] Apply symmetric change to python3.rb

tdsmith commented:
Ping @sjackman for review; this works, but I don't know much about your preferred style for Linuxbrew changes. Let me know if anything needs tweaking.

tseemann commented:
@tdsmith @sjackman  

 I can confirm that your modified recipe works on RHEL 7.x for python 2 and 3 (Hooray!)

sjackman commented:
Thanks for this, Tim. Does compiling Python work out-of-the-box using a Homebrewed `gcc`? Does this patch support primarily using the system's `/usr/bin/gcc`?

sjackman commented on Library/Formula/python.rb:
```diff
@@ -106,6 +106,17 @@ def install
+      hint_libs = hints.map { |f| "-L#{Formula[f].opt_lib}" }
+      cppflags = "CPPFLAGS=#{hint_includes.join " "}"
+      ldflags = "LDFLAGS=#{hint_libs.join " "}"
+      args << cppflags << ldflags
```
Does `-I#{HOMEBREW_PREFIX}/include -L#{HOMEBREW_PREFIX}/lib` work?

tdsmith replied:
It may; I hadn't appreciated that these formulas aren't keg-only on Linux. I'll try that later.

sjackman commented on Library/Formula/python.rb:
```diff
@@ -115,7 +126,7 @@ def install
     inreplace "setup.py" do |s|
       s.gsub! "do_readline = self.compiler.find_library_file(lib_dirs, 'readline')",
-              "do_readline = '#{Formula["readline"].opt_lib}/libhistory.dylib'"
+              "do_readline = '#{Formula["readline"].opt_lib}/libhistory.so'"
```
Do you get an error message without this fix? What is the effect of leaving this as is?

tdsmith replied:
No, but Python fails to detect whether readline is linked against curses or not, which can cause the _curses module to misbuild.

tdsmith replied:
Fixed the style.

sjackman commented:
> I can confirm that your modified recipe works on RHEL 7.x for python 2 and 3

Good to know! I'm very curious as to why though, because Homebrew already adds to the command line
`-I#{HOMEBREW_PREFIX}/include -L#{HOMEBREW_PREFIX}/lib`

sjackman commented:
Oh, I think I see now. It looks like it's because the `python` formula is completely overriding the standard Homebrew definition of `CPPFLAGS`, `CFLAGS` and `LDFLAGS`. Is that correct?

tdsmith commented:
I just saw your comment about errors, Torsten; I'm guilty of this too, but can you add a new comment instead of editing when the content substantially changes? A new comment triggers an email notification but an edit doesn't.

> I can't seem to pip install things that need to compile eg. numpy (needs -dev stuff) and also it seems to be using system BLAS. Is this why I'm mean to use homebrew-python/numpy ?

Yes. Our Python configuration exposes a bug in numpy's custom distutils that we need to patch around that yields that error message about headers. The fix has been accepted upstream but not released. :( It's only relevant when building from source, which most OS X users don't do because pypi hosts wheels. As far as which BLAS is used, that's something you have to configure at numpy's build time, which the Homebrew formula takes care of.

> Does compiling Python work out-of-the-box using a Homebrewed gcc?

I haven't tested.

> Does this patch support primarily using the system's /usr/bin/gcc?

Yes.

> Oh, I think I see now. It looks like it's because the python formula is completely overriding the standard Homebrew definition of CPPFLAGS, CFLAGS and LDFLAGS. Is that correct?

It may actually be doing that now but it was not doing that before; Python's hacky setup.py, which builds the compiled extensions, parses the FLAGS variables to learn search paths for the dependencies of the compiled extension modules.

> Homebrew already adds to the command line `-I#{HOMEBREW_PREFIX}/include`

If you mean literally that Homebrew is wrapping compiler invocations and appending this to the argument list, that won't help, because setup.py places the discovered include path earlier on the command line.

sjackman commented:
>> Homebrew already adds to the command line -I#{HOMEBREW_PREFIX}/include

> If you mean literally that Homebrew is wrapping compiler invocations and appending this to the argument list, that won't help, because setup.py places the discovered include path earlier on the command line.

Homebrew adds the `-I#{HOMEBREW_PREFIX}/include` option to `CPPFLAGS` and `-L#{HOMEBREW_PREFIX}/lib` to `LDFLAGS`.

Linuxbrew's `gcc` adds `-L#{HOMEBREW_PREFIX}/lib` to the [`specs` file](https://github.com/Homebrew/linuxbrew/blob/master/Library/Formula/gcc.rb#L266-L270), which precedes the values specified on the command line. I only just noticed that it does *not* add `-I#{HOMEBREW_PREFIX}/include` to the `specs` file. It really should! That may be the ultimate source of a lot of our Python woes.

sjackman commented:
> because setup.py places the discovered include path earlier on the command line.

That's terrible.

tdsmith commented:
Ah: Homebrew uses `-isystem`, not `-I`.
```
[vagrant@localhost Python-2.7.10]$ echo $CPPFLAGS
-isystem/home/vagrant/.linuxbrew/include
```
setup.py ignores `-i`. Adding `-I` is sufficient.

Updated to just hint the Linuxbrew prefix's include and library paths. Also added a fix for building the _curses module, which I noticed was failing.

FWIW, I've received some strenuous feedback that prioritizing system search paths ahead of user-specified search paths is undesirable, e.g. here: https://github.com/Homebrew/homebrew/issues/26272#issuecomment-96406197 -- it's hard to accommodate that in Python in stdenv without fixing up the memorized configuration variables.

sjackman commented:
@zpincus wrote https://github.com/Homebrew/homebrew/issues/26272#issuecomment-96497263
> if that 3rd-party module contains a C extension that has a private header with the same name as any header file that is in /usr/local/include (or, more generally, brew --prefix). The same is true for linking libraries.

I've had this same issue outside of Python. See https://github.com/Homebrew/homebrew-science/issues/2458#issuecomment-117318345

sjackman commented:
The challenge is that we want to prioritize and enforce the following:
1. the package's private header/library files
2. Homebrew's header/library files
3. the system's header library files
4. and never the user's non-system header library/files

It's difficult to get 1 before 2 while still getting 2 before 3.

prologic commented:
This worked for me too. I also commented on #420 but I believe this PR fixes the issue at hand.

prologic commented:
NB: I hand patched my ``python.rb`` formula however since this PR containers merge conflicts apparently

sjackman commented:
Hi, Tim. Sorry for the long delay in getting around to reviewing your patch. I've applied a reduced version of this patch in cd680d7 that addresses the core issue, which is adding `-I#{HOMEBREW_PREFIX}/include` to `CPPFLAGS`. Thanks for helping troubleshoot this issue.

I haven't seen any issues filed regarding `ncurses` for Python, so I've left that chunk out for now.
